config/config_test.go

```diff
@@ -175,4 +177,48 @@ var _ = Describe("Config", func() {
 			Expect(cfg.CaCerts).To(HaveLen(0))
 		})
 	})
+
+	Describe("#Client", func() {
+		Context("when client credentials are supplied", func() {
+			It("uses the OAuth auth strategy", func() {
+				cfg.ClientID = "a-client-id"
+				cfg.ClientSecret = "a-client-secret"
+
+				client, err := cfg.Client()
+				Expect(err).ToNot(HaveOccurred())
+				Expect(client.Auth).To(BeAssignableToTypeOf(&auth.OAuthStrategy{}))
+			})
+		})
+
+		Context("when non-client credentials are supplied", func() {
```

Comment:
Honestly I don't really understand how the auth works when `clientId` and `clientSecret` are empty. That's why this test is awkwardly named. I don't suppose you understand the `useClientCredentials = false` functionality of `credhub.Auth`?

Reply:
OK so I agree that it is confusing. As far as I understand, with the CredHub CLI, as of now, there are only 2 ways of logging in: client credentials or password grant.
1. When you use the client credentials grant type, you are required to pass a client name and a client secret, and you log in with these. (that is quite normal)
1. When you use the password grant type, which theoretically requires 4 things: username, password, client name, client secret; you _only_ can pass the username and password, and the CLI uses its hardcoded `cf` client with empty password to log you in. So when you *don't* pass the client name and secret, you are using the password grant flow.


> I don't suppose you understand the useClientCredentials = false functionality of credhub.Auth

 `useClientCredentials` is set to false to indicate that we are not using client credentials (hence we are using password grant,  since these are the 2 only supported methods), which appears to be used [downstream](https://github.com/cloudfoundry-incubator/credhub-cli/blob/b37fc2b1557c52124b604da2dd0caeff7c89ae1c/credhub/auth/oauth.go#L146-L150) to decide whether to do a client credentials or password grant when making the OAuth call to the UAA.

To sum-up, it sounds like what you want is to have 3 different cases for your test:
* client credentials grant (OAuth)
* password grant (OAuth) 
* the new mTLS method

So maybe this context can be called "when password credentials are supplied", or maybe "when a password grant is attempted"...

Please let us know if that's not clear :)